apiserver/internal/handlers/objects/objects.go

```diff
@@ -8,391 +8,102 @@ import (
-		Architecture: architecture,
-	}, nil
-}
-
-// CharmURLFromLocatorDuringMigration returns a charm URL from a charm locator
-// during model migration.
-// By including the architecture only when it's passed in, will allow us to
-// move forward with prior versions (3.x) not passing in the architecture and
-// current versions (4.x) passing in the architecture.
-func CharmURLFromLocatorDuringMigration(locator applicationcharm.CharmLocator, includeArchitecture bool) (*charm.URL, error) {
-	schema, err := convertSource(locator.Source)
-	if err != nil {
-		return nil, errors.Capture(err)
-	}
-
-	var architecture string
-	if includeArchitecture {
-		architecture, err = convertApplication(locator.Architecture)
-		if err != nil {
-			return nil, errors.Capture(err)
-		}
-	}
-
-	return &charm.URL{
-		Schema:       schema,
-		Name:         locator.Name,
-		Revision:     locator.Revision,
-		Architecture: architecture,
-	}, nil
-}
-
-func splitNameAndSHAFromQuery(query url.Values) (string, string, error) {
-	charmObjectID := query.Get(":object")
-
-	// Path param is {charmName}-{charmSha256[0:7]} so we need to split it.
-	// NOTE: charmName can contain "-", so we cannot simply strings.Split
-	splitIndex := strings.LastIndex(charmObjectID, "-")
-	if splitIndex == -1 {
-		return "", "", jujuerrors.BadRequestf("%q is not a valid charm object path", charmObjectID)
-	}
-	name, sha := charmObjectID[:splitIndex], charmObjectID[splitIndex+1:]
-	if len(sha) < corecharm.MinSHA256PrefixLength {
-		return "", "", jujuerrors.BadRequestf("invalid sha length: %q", sha)
-	}
-	return name, sha, nil
-}
-
-// sendJSONError sends a JSON-encoded error response.  Note the
-// difference from the error response sent by the sendError function -
-// the error is encoded in the Error field as a string, not an Error
-// object.
-func sendJSONError(w http.ResponseWriter, err error) error {
-	perr, status := apiservererrors.ServerErrorAndStatus(err)
-	return errors.Capture(internalhttp.SendStatusAndJSON(w, status, &params.CharmsResponse{
-		Error:     perr.Message,
-		ErrorCode: perr.Code,
-		ErrorInfo: perr.Info,
-	}))
-}
-
-// ModelState is an interface for getting the model state.
-type ModelState interface {
-	Model() (Model, error)
-	Release() bool
-}
-
-// Model is an interface for getting the model migration mode.
-type Model interface {
-	MigrationMode() state.MigrationMode
-}
-
-func modelIsImporting(st ModelState) (bool, error) {
-	model, err := st.Model()
-	if err != nil {
-		return false, errors.Capture(err)
-	}
-	return model.MigrationMode() == state.MigrationModeImporting, nil
-}
-
-func convertSource(source applicationcharm.CharmSource) (string, error) {
-	switch source {
-	case applicationcharm.CharmHubSource:
-		return charm.CharmHub.String(), nil
-	case applicationcharm.LocalSource:
-		return charm.Local.String(), nil
-	default:
-		return "", errors.Errorf("unsupported source %q", source)
+		return errors.Errorf("error processing charm archive download: %w", err)
```

Comment:
If this is a generic object handler then we shouldn't have any references to `charm` here.

Reply:
Good shout.